app/services/articles/feeds/weighted_query_strategy.rb

```diff
@@ -0,0 +1,542 @@
+            ON comments.commentable_id = articles.id
+              AND comments.commentable_type = 'Article'
+          WHERE #{where_clause}
+          GROUP BY articles.id
+          ORDER BY (#{relevance_score_components_as_sql}) DESC,
+            articles.published_at DESC
+          #{offset_and_limit_clause(offset: offset, limit: limit)}
+        THE_SQL_STATEMENT
+      end
+
+      # The sql statement for selecting based on relevance scores that
+      # are user required.
+      def sql_sub_query_for_existing_user(only_featured:, must_have_main_image:, limit:, offset:, omit_article_ids:)
+        where_clause = build_sql_with_where_clauses(
+          only_featured: only_featured,
+          must_have_main_image: must_have_main_image,
+          omit_article_ids: omit_article_ids,
+        )
+        <<~THE_SQL_STATEMENT
+          SELECT articles.id
+          FROM articles
+          LEFT OUTER JOIN taggings
+            ON taggings.taggable_id = articles.id
+              AND taggable_type = 'Article'
+          INNER JOIN tags
+            ON taggings.tag_id = tags.id
+          LEFT OUTER JOIN follows AS followed_tags
+            ON tags.id = followed_tags.followable_id
+              AND followed_tags.followable_type = 'ActsAsTaggableOn::Tag'
+              AND followed_tags.follower_type = 'User'
+              AND followed_tags.follower_id = :user_id
+              AND followed_tags.explicit_points >= 0
+          LEFT OUTER JOIN follows AS followed_user
+            ON articles.user_id = followed_user.followable_id
+              AND followed_user.followable_type = 'User'
+              AND followed_user.follower_id = :user_id
+              AND followed_user.follower_type = 'User'
+          LEFT OUTER JOIN follows AS followed_org
+            ON articles.organization_id = followed_org.followable_id
+              AND followed_org.followable_type = 'Organization'
+              AND followed_org.follower_id = :user_id
+              AND followed_org.follower_type = 'User'
+          LEFT OUTER JOIN comments AS comments_by_followed
+            ON comments_by_followed.commentable_id = articles.id
+              AND comments_by_followed.commentable_type = 'Article'
+              AND followed_user.followable_id = comments_by_followed.user_id
+              AND followed_user.followable_type = 'User'
+          LEFT OUTER JOIN user_blocks
+            ON user_blocks.blocked_id = articles.user_id
+              AND user_blocks.blocked_id IS NULL
+              AND user_blocks.blocker_id = :user_id
+          LEFT OUTER JOIN comments
+            ON comments.commentable_id = articles.id
+              AND comments.commentable_type = 'Article'
+          WHERE #{where_clause}
+          GROUP BY articles.id
+          ORDER BY (#{relevance_score_components_as_sql}) DESC,
+            articles.published_at DESC
+            #{offset_and_limit_clause(offset: offset, limit: limit)}
+        THE_SQL_STATEMENT
```

Comment:
Question: This is a very general question, but how does this query perform in production? Do we have ALL the correct indices present?

Reply:
I've run the query in Blazer and ran describe on the query, and things appear to hit indices.  But I am up for some help.  I'll post the resulting query.

```sql
SELECT "articles"."path", "articles"."title", "articles"."id", "articles"."published", "articles"."comments_count", "articles"."public_reactions_count", "articles"."cached_tag_list", "articles"."main_image", "articles"."main_image_background_hex_color", "articles"."updated_at", "articles"."slug", "articles"."video", "articles"."user_id", "articles"."organization_id", "articles"."video_source_url", "articles"."video_code", "articles"."video_thumbnail_url", "articles"."video_closed_caption_track_url", "articles"."experience_level_rating", "articles"."experience_level_rating_distribution", "articles"."cached_user", "articles"."cached_organization", "articles"."published_at", "articles"."crossposted_at", "articles"."description", "articles"."reading_time", "articles"."video_duration_in_seconds", "articles"."last_comment_at" FROM "articles" WHERE "articles"."id" IN (SELECT articles.id
FROM articles
LEFT OUTER JOIN taggings
  ON taggings.taggable_id = articles.id
    AND taggable_type = 'Article'
INNER JOIN tags
  ON taggings.tag_id = tags.id
LEFT OUTER JOIN follows AS followed_tags
  ON tags.id = followed_tags.followable_id
    AND followed_tags.followable_type = 'ActsAsTaggableOn::Tag'
    AND followed_tags.follower_type = 'User'
    AND followed_tags.follower_id = 111
    AND followed_tags.explicit_points >= 0
LEFT OUTER JOIN follows AS followed_user
  ON articles.user_id = followed_user.followable_id
    AND followed_user.followable_type = 'User'
    AND followed_user.follower_id = 111
    AND followed_user.follower_type = 'User'
LEFT OUTER JOIN follows AS followed_org
  ON articles.organization_id = followed_org.followable_id
    AND followed_org.followable_type = 'Organization'
    AND followed_org.follower_id = 111
    AND followed_org.follower_type = 'User'
LEFT OUTER JOIN comments AS comments_by_followed
  ON comments_by_followed.commentable_id = articles.id
    AND comments_by_followed.commentable_type = 'Article'
    AND followed_user.followable_id = comments_by_followed.user_id
    AND followed_user.followable_type = 'User'
LEFT OUTER JOIN user_blocks
  ON user_blocks.blocked_id = articles.user_id
    AND user_blocks.blocked_id IS NULL
    AND user_blocks.blocker_id = 111
LEFT OUTER JOIN comments
  ON comments.commentable_id = articles.id
    AND comments.commentable_type = 'Article'
WHERE articles.published = true AND articles.published_at > '2021-10-15 22:50:37.364177' AND articles.published_at < '2021-11-15 23:50:37.364460'
GROUP BY articles.id
ORDER BY ((CASE (current_date - published_at::date)
WHEN 0 THEN 1.0
WHEN 1 THEN 0.95
WHEN 2 THEN 0.9
WHEN 3 THEN 0.85
WHEN 4 THEN 0.8
WHEN 5 THEN 0.75
WHEN 6 THEN 0.7
WHEN 7 THEN 0.65
WHEN 8 THEN 0.6
WHEN 9 THEN 0.55
WHEN 10 THEN 0.5
WHEN 11 THEN 0.4
WHEN 12 THEN 0.3
WHEN 13 THEN 0.2
WHEN 14 THEN 0.1
ELSE 0.001 END) * (CASE COUNT(comments_by_followed.id)
WHEN 0 THEN 0.95
WHEN 1 THEN 0.98
WHEN 2 THEN 0.99
ELSE 0.93 END) * (CASE articles.comments_count
WHEN 0 THEN 0.9
WHEN 1 THEN 0.94
WHEN 2 THEN 0.95
WHEN 3 THEN 0.98
WHEN 4 THEN 0.999
ELSE 1.0 END) * (CASE ROUND(ABS(articles.experience_level_rating - (SELECT
              (CASE
                 WHEN experience_level IS NULL THEN 5
                 ELSE experience_level END ) AS user_experience_level
              FROM users_settings WHERE users_settings.user_id = 111
            )))
WHEN 0 THEN 1.0
WHEN 1 THEN 0.98
WHEN 2 THEN 0.97
WHEN 3 THEN 0.96
WHEN 4 THEN 0.95
WHEN 5 THEN 0.94
ELSE 0.93 END) * (CASE (CASE articles.featured WHEN true THEN 1 ELSE 0 END)
WHEN 1 THEN 1.0
ELSE 0.85 END) * (CASE COUNT(followed_user.follower_id)
WHEN 0 THEN 0.8
WHEN 1 THEN 1.0
ELSE 1.0 END) * (CASE COUNT(followed_org.follower_id)
WHEN 0 THEN 0.95
WHEN 1 THEN 1.0
ELSE 1.0 END) * (CASE (current_date - MAX(comments.created_at)::date)
WHEN 0 THEN 1.0
WHEN 1 THEN 0.9988
ELSE 0.988 END) * (CASE COUNT(followed_tags.follower_id)
WHEN 0 THEN 0.4
WHEN 1 THEN 0.9
ELSE 1.0 END) * (CASE (CASE
                 WHEN articles.privileged_users_reaction_points_sum < -10 THEN -1
                 WHEN articles.privileged_users_reaction_points_sum > 10 THEN 1
                 ELSE 0 END)
WHEN -1 THEN 0.2
WHEN 1 THEN 1.0
ELSE 0.95 END) * (CASE articles.reactions_count
WHEN 0 THEN 0.9988
WHEN 1 THEN 0.9988
WHEN 2 THEN 0.9988
WHEN 3 THEN 0.9988
ELSE 1.0 END) * (CASE articles.spaminess_rating
WHEN 0 THEN 1.0
ELSE 0.0 END)) DESC,
  articles.published_at DESC
  LIMIT 50
) ORDER BY "articles"."published_at" DESC
```

Paging @jgaskins to give the ol' "Check the index."

Reply:
Here's the analysis results:

Node type:	Count

* Aggregate: 1
* Hash: 8
* Hash Inner Join: 3
* Hash Left Join: 5
* Index Scan: 4
* Limi: 1
* Seq Scan: 6
* Sort: 2
* Subquery Scan: 1

Relation name	Scan count
Node type	Count

* articles: 2
* Seq Scan: 2
* comments: 2
* Seq Scan: 2
* follows: 3
* Index Scan: 3
* taggings: 1
* Seq Scan: 1
* tags: 1
* Seq Scan: 1
* users_settings: 1
* Index Scan: 1

Reply:
this works [locally](http://localhost:3000/admin/blazer) just fine, but DEV's db (older pg version?) raises an error from the query in the comment: 

> column "articles.published_at" must appear in the GROUP BY clause or be used in an aggregate function LINE 38: ORDER BY ((CASE (current_date - published_at::date)

Reply:
There are a few more `GROUP BY` that I need to add (or we upgrade Postgres?)

Reply:
@djuber I have an updated query to test.  I ran it in Blazer:

```sql
SELECT "articles"."path", "articles"."title", "articles"."id", "articles"."published", "articles"."comments_count", "articles"."public_reactions_count", "articles"."cached_tag_list", "articles"."main_image", "articles"."main_image_background_hex_color", "articles"."updated_at", "articles"."slug", "articles"."video", "articles"."user_id", "articles"."organization_id", "articles"."video_source_url", "articles"."video_code", "articles"."video_thumbnail_url", "articles"."video_closed_caption_track_url", "articles"."experience_level_rating", "articles"."experience_level_rating_distribution", "articles"."cached_user", "articles"."cached_organization", "articles"."published_at", "articles"."crossposted_at", "articles"."description", "articles"."reading_time", "articles"."video_duration_in_seconds", "articles"."last_comment_at" FROM "articles" WHERE "articles"."id" IN (SELECT articles.id
FROM articles
LEFT OUTER JOIN taggings
  ON taggings.taggable_id = articles.id
    AND taggable_type = 'Article'
INNER JOIN tags
  ON taggings.tag_id = tags.id
LEFT OUTER JOIN follows AS followed_tags
  ON tags.id = followed_tags.followable_id
    AND followed_tags.followable_type = 'ActsAsTaggableOn::Tag'
    AND followed_tags.follower_type = 'User'
    AND followed_tags.follower_id = 12
    AND followed_tags.explicit_points >= 0
LEFT OUTER JOIN follows AS followed_user
  ON articles.user_id = followed_user.followable_id
    AND followed_user.followable_type = 'User'
    AND followed_user.follower_id = 12
    AND followed_user.follower_type = 'User'
LEFT OUTER JOIN follows AS followed_org
  ON articles.organization_id = followed_org.followable_id
    AND followed_org.followable_type = 'Organization'
    AND followed_org.follower_id = 12
    AND followed_org.follower_type = 'User'
LEFT OUTER JOIN comments AS comments_by_followed
  ON comments_by_followed.commentable_id = articles.id
    AND comments_by_followed.commentable_type = 'Article'
    AND followed_user.followable_id = comments_by_followed.user_id
    AND followed_user.followable_type = 'User'
LEFT OUTER JOIN user_blocks
  ON user_blocks.blocked_id = articles.user_id
    AND user_blocks.blocked_id IS NULL
    AND user_blocks.blocker_id = 12
LEFT OUTER JOIN comments
  ON comments.commentable_id = articles.id
    AND comments.commentable_type = 'Article'
WHERE articles.published = true AND articles.published_at > '2021-10-16 15:48:00.006806' AND articles.published_at < '2021-11-16 16:48:00.007201'
GROUP BY articles.id, articles.published_at, articles.comments_count, articles.experience_level_rating, articles.featured, articles.privileged_users_reaction_points_sum, articles.reactions_count, articles.spaminess_rating
ORDER BY ((CASE (current_date - articles.published_at::date)
WHEN 0 THEN 1.0
WHEN 1 THEN 0.95
WHEN 2 THEN 0.9
WHEN 3 THEN 0.85
WHEN 4 THEN 0.8
WHEN 5 THEN 0.75
WHEN 6 THEN 0.7
WHEN 7 THEN 0.65
WHEN 8 THEN 0.6
WHEN 9 THEN 0.55
WHEN 10 THEN 0.5
WHEN 11 THEN 0.4
WHEN 12 THEN 0.3
WHEN 13 THEN 0.2
WHEN 14 THEN 0.1
ELSE 0.001 END) *
(CASE COUNT(comments_by_followed.id)
WHEN 0 THEN 0.95
WHEN 1 THEN 0.98
WHEN 2 THEN 0.99
ELSE 0.93 END) *
(CASE articles.comments_count
WHEN 0 THEN 0.9
WHEN 1 THEN 0.94
WHEN 2 THEN 0.95
WHEN 3 THEN 0.98
WHEN 4 THEN 0.999
ELSE 1.0 END) *
(CASE ROUND(ABS(articles.experience_level_rating - (SELECT
              (CASE
                 WHEN experience_level IS NULL THEN 5
                 ELSE experience_level END ) AS user_experience_level
              FROM users_settings WHERE users_settings.user_id = 12
            )))
WHEN 0 THEN 1.0
WHEN 1 THEN 0.98
WHEN 2 THEN 0.97
WHEN 3 THEN 0.96
WHEN 4 THEN 0.95
WHEN 5 THEN 0.94
ELSE 0.93 END) *
(CASE (CASE articles.featured WHEN true THEN 1 ELSE 0 END)
WHEN 1 THEN 1.0
ELSE 0.85 END) *
(CASE COUNT(followed_user.follower_id)
WHEN 0 THEN 0.8
WHEN 1 THEN 1.0
ELSE 1.0 END) *
(CASE COUNT(followed_org.follower_id)
WHEN 0 THEN 0.95
WHEN 1 THEN 1.0
ELSE 1.0 END) *
(CASE (current_date - MAX(comments.created_at)::date)
WHEN 0 THEN 1.0
WHEN 1 THEN 0.9988
ELSE 0.988 END) *
(CASE COUNT(followed_tags.follower_id)
WHEN 0 THEN 0.4
WHEN 1 THEN 0.9
ELSE 1.0 END) *
(CASE (CASE
                 WHEN articles.privileged_users_reaction_points_sum < -10 THEN -1
                 WHEN articles.privileged_users_reaction_points_sum > 10 THEN 1
                 ELSE 0 END)
WHEN -1 THEN 0.2
WHEN 1 THEN 1.0
ELSE 0.95 END) *
(CASE articles.reactions_count
WHEN 0 THEN 0.9988
WHEN 1 THEN 0.9988
WHEN 2 THEN 0.9988
WHEN 3 THEN 0.9988
ELSE 1.0 END) *
(CASE articles.spaminess_rating
WHEN 0 THEN 1.0
ELSE 0.0 END)) DESC,
  articles.published_at DESC
  LIMIT 50
) ORDER BY "articles"."published_at" DESC
```

Reply:
yeah - sorry for the confusion as I was actively expanding my earlier comment - my results, after doing some checking, is this is not a postgresql version issue (first hypothesis) or a permissions issue (second hypothesis) but was a problem running a query against a view (reproducible on simple cases regardless of permissions).

your update (adding more articles columns to the group by, not just `id`) does allow the query to run against the view correctly.

Reply:
Excellent, glad we could figure this out and resolve it.

@djuber are you in a position to answer the question about indexing?  That's the major and unaddressed concern that @joshpuetz and I have.

Reply:
I'm not sure how helpful it is, but I ran the above query on production via Heroku Dataclips and it executes quite quickly:

https://data.heroku.com/dataclips/bazyxojrjhxahwbbcabrucocgiwx

Reply:
I have a suspicion that this change will be more SQL performant than it's competitor.  The competitor has a 2N+1 situation as it parsed the result set and then runs queries on each result.

Whereas this pushes that work into the SQL statement.

Reply:
@jeremyf A few 4am notes, because sleep doesn't appear to be happening this evening 😄 These are just notes. I haven't looked at it enough for any of it to be a critique. In fact, I may also be duplicating thoughts others have already provided. This is very much just an episode of Jamie's insomniac ramblings.

- Both versions of this query execute on DEV's production DB in 350-360ms measured by the DB server itself (rather than from the perspective of a client, which would count network latency)
- <details><summary>p90 for the homepage at the same time and query volume is currently 222ms</summary><a href="https://app.datadoghq.com/apm/resource/practical_developer/rack.request/29bdfb87f415d6a4?query=env%3Aproduction%20service%3Apractical_developer%20operation_name%3Arack.request%20resource_name%3AStoriesController%23index&env=production&index=apm-search&topGraphs=latency%3Alatency%2CbreakdownAs%3Apercentage%2Cerrors%3Acount%2Chits%3Acount&start=1637053564314&end=1637139964314&paused=false">Datadog link</a><br/><img width="201" alt="Screenshot of percentile latency for the StoriesController index action. p99 is 658ms, p90 is 222ms, and p50 is 131ms" src="https://user-images.githubusercontent.com/108205/142170040-323aabb4-742f-4d9b-8277-ac33597db1e1.png"></summary>
- `JOIN` vs N queries is typically a tradeoff of network latency vs DBMS resources
  - Per-resource-type queries (via ActiveRecord's `includes` — effectively breadth-first graph traversal) typically query against either PK or FK indexes and raw tables
  - `JOIN`s require transforming raw tabular data into a projection of all relations being queried, rather than 
  - The hard part is that we don't have a lot of visibility into how we're using DBMS resources on DEV since it's still running on a Heroku Postgres instance, which does not provide CPU/RAM metrics in a format that is understood by Datadog.
- The query plan indicates a pretty heavy query
  - <details><summary>the data structure used by <code>LEFT JOIN comments</code> uses 42MB of RAM on its own</summary>
  
    ```
    ->  Parallel Hash  (cost=69201.38..69201.38 rows=178947 width=16) (actual time=191.666..191.667 rows=179006 loops=4)
     Buckets: 1048576  Batches: 1  Memory Usage: 41856kB
     ->  Parallel Seq Scan on comments  (cost=0.00..69201.38 rows=178947 width=16) (actual time=0.022..132.941 rows=179006 loops=4)
           Filter: ((commentable_type)::text = 'Article'::text)
           Rows Removed by Filter: 3957
    ```

    </details>
  - It also appears to be taking up 192 of the 350-360 total ms used by this query
  - The fact that it's doing a sequential scan to filter on `commentable_type` shouldn't be impacting it much since it's only removing a little under 4k rows
- The query plan is executed based on machine heuristics in addition to data heuristics
  - On a DB server with less memory available (such as a self-hosted Forem instance), the `Parallel Hash` join above may instead be run as a `Nested Loop Left Join`
  - A really handsome engineer wrote [a blog post about Postgres query-planner shenanigans](https://dev.to/jgaskins/tricksy-little-postgres-query-planner-42hg) like this
  - Some of the `Parallel` parts of the query plan may also not be done in parallel if the CPU is under heavy load
- I'd also be curious what the impact is on a Forem with a lower volume of traffic than DEV but still significant volume of data, such as CodeNewbie, but I haven't checked

Again, these are just some musings from looking at the query plan and cross-referencing with Datadog after a night without sleep, so I may very well be misreading any number of things in it and I'd be happy for anyone to check me on any of it. I'm also happy to talk more about it after I've gotten some sleep.

Reply:
@jgaskins My question for you…do you think we can deploy this and optimize later?  Or do we need to go back and rework the queries?

If we need to go back and rework, I'm going to need some assistance.  (I think I can do it if I narrow my initial scope and rethink my approach, but someone that's went down the path would be helpful).

Again, we do have the "And if this goes south, we can update the application ENV to mitigate the problem."

Reply:
> My question for you…do you think we can deploy this and optimize later? Or do we need to go back and rework the queries?
>
> Again, we do have the "And if this goes south, we can update the application ENV to mitigate the problem."

@jeremyf Short answer: It looks like we can deploy. You added [a kill switch](https://github.com/forem/forem/pull/15240#discussion_r743884178) for if things get too hot, and I'm down with that. Worst-case scenario: we roll back the deployment and disable deployment until a fix is merged. If we're missing indexes, we can put in temporary ones on DEV while the real ones are written and pushed through CI. Predicting index usage is hard, especially on a query this complicated.

Slightly longer answer: For the kill switches, normally we use [feature flags](https://developers.forem.com/technical-overview/feature-flags)<sup>1</sup>, backed by [the `flipper` gem](https://github.com/jnunemaker/flipper), but the solution you mentioned a couple weeks ago should also work fine. I would like to get into the habit of using them more to decouple release from deployment, but that requires a bit more effort and we're not there yet.

We can also build in circuit breakers that, when tripped a certain number of times/percentage of the time, disable the feature flag automatically. I have an implementation of that locally, but the idea is that we would use that for anything we consider risky to avoid requiring a human response. Basically: automating response to [SLOs](https://sre.google/sre-book/service-level-objectives/#objectives-g0s1tdcz) on feature flags.

<sup>1</sup> I apparently didn't mention that when I reviewed this before. Sorry about that, it probably got lost in editing. Feature flags are [available to nearly everyone with admin access](https://dev.to/admin/feature_flags/features) and are live in 30 seconds on DEV. Normally feature flags would be instant, but we [cache them for 5 seconds by default](https://github.com/forem/forem/pull/14668) and DEV makes it 30 to reduce DB pressure even further. ENV var changes on Heroku [can take 3 minutes](https://devcenter.heroku.com/articles/preboot#deploying-with-preboot), but we have enough capacity to handle 3 minutes of non-ideal performance. The nu